### ISSUE TITLE
API-incompatible change in 3.15.0 caused by adding parallel_direct_writes to bit field in fuse_file_info

### ISSUE THREAD

**Issue by mxmlnkn** [libfuse/libfuse#1029]
Hello,

while looking over the `fuse_file_info` definition in `fuse_common.h` in order to add FUSE 3 support to fusepy, I saw that libfuse 3.14.0 introduced a new bitfield member `parallel_direct_writes` in https://github.com/libfuse/libfuse/commit/a5eb7f2a0117ab43119ef5724cf5f4f2f181804a in pull request #659. I'm thinking that this is a major bug because it is an incompatible change. Programs built against libfuse 3.14 and using a shared library of libfuse 3.15, or built against 3.15 and getting a shared libfuse < 3.15, will fail in intricate ways because all bit fields after the newly introduced member will be off by one. This affects: `flush`, `nonseekable`, `flock_release`, `cache_readdir`, `noflush`.

If this is indeed a bug, I can do a pull request that moves `parallel_direct_writes` to right before the `padding` member.

**Comment by maksim-harbachou-resilio** [comment]
The same PR changed `struct fuse_config` layout. If I'm not missing something, `struct fuse_config` is a part of public interface. As a result, any code built with different libfuse3 headers will pass config object that is not valid for another (newer / older) libfuse3.so. I think it's a major issue.

**Comment by bsbernd** [comment]
@mxmlnkn yes it is bug, the problem is that there had been releases in between - we have the choice to break ABI for < 3.14.1 and anything after 3.14.1. 

Also see here  https://lists.debian.org/debian-devel/2024/03/msg00278.html

Which is why current master branch has API changes to compile in the library version to be able to better fix it on the fly if it should ever happen again.

I have a patch to fix fields based on heuristics at run time, but I simply didn't come to it to test it yet. For the current API change, which includes the fuse version, I also need to check for ABI compatibility - these things are currently blocking the release of 3.17 (already ways too long).

**Comment by bsbernd** [comment]
@maksim-harbachou-resilio Yes, fuse_config got also changed, but it is also got broken a couple of times before. The comment at the end of the structure was rather confusing and is already updated in master.

**Comment by mxmlnkn** [comment]
I think adding the `no_rofd_flush` member to `fuse_config` in #635, which landed in libfuse 3.11 was also an ABI-breaking change. The linked Debian thread does not seem to discuss this member, only `parallel_direct_writes` added in 3.14.1, a bugfix release! I also don't see any issues mentioning `no_rofd_flush`.

> the problem is that there had been releases in between - we have the choice to break ABI for < 3.14.1 and anything after 3.14.1.

I don't think, I can decide that. Personally, I feel like it is a bug and should be changed back, but seeing that there are one, possibly two, other ABI-breaking versions out there already, I guess simply leaving it as is and documenting it, could also be fine? At least, I can iron out some of these kinks with fusepy acting as an intermediate as long as I know about them and can check against a minor version.

If it is reverted, then it should be done in a minor release, not as a bugfix, because there does not even seem to be a nice API for querying the bugfix version. `get_fuse_version` only returns `major * 100 + minor` since 3.10 and `major * 10 + minor)` before, which I also found an unexpected API change that might have broken some version checking code, I'd guess. It should be possible to parse the output of `fuse_main` called with `--version` to get the buffix version, but then I would have to redirect stdout and parse text.

**Comment by bsbernd** [comment]
Yeah, I know about no_rofd_flush, I had added it in the private discussion we initially had. I found it through this site:

https://abi-laboratory.pro/index.php?view=timeline&l=libfuse

Looks like that site doesn't get updated anymore (I hope the developer is ok, the software for that is here: 
https://abi-laboratory.pro/index.php?view=timeline&l=libfuse
I want to use it in ci tests, although I'm not sure if I can fully trust automatic testing. Maybe I need to sometimes override it - it reports a weird issue in 3.12, but I can't make anything out of it yet
https://abi-laboratory.pro/index.php?view=compat_report&l=libfuse&v1=3.11.0&v2=3.12.0&obj=0a814&kind=abi#Symbol_Problems_High

`The parameter se became passed in r12 register instead of rbp`

(Well, there is an issue with 3.12, if you use the wrong FUSE_USE_VERSION and the new API).

Personally I don't feel well about correcting the parallel-dio commit - either documenting or doing heuristics, or both is the way to go. Here is what I have for the heuristics so far (totally untested)
https://github.com/bsbernd/libfuse/commit/2aba73e8bb9e0ed39ff4208ac5b99967b053f5d5
The other way would be to correct and then to do the heuristics the other way around.

**Comment by bsbernd** [comment]
I'm closing it, we have restored the 3.10 ABI, but also bumped the .so version.
